### PR TITLE
Upgrade to sbt 1

### PR DESCRIPTION
**Pull request by Duhemm**
Projects from the community build to port to sbt1:

- [x] [algebra](https://github.com/dotty-staging/algebra/tree/sbt1)
- [x] [better-files](https://github.com/dotty-staging/better-files/tree/sbt1)
- [x] [scalacheck](https://github.com/dotty-staging/scalacheck/tree/sbt1)
- [x] [ScalaPB](https://github.com/dotty-staging/ScalaPB/tree/sbt1)
- [x] [scalatest](https://github.com/dotty-staging/scalatest/tree/sbt1)
- [x] [scopt](https://github.com/dotty-staging/scopt/tree/sbt1)
- [x] [squants](https://github.com/dotty-staging/squants/tree/sbt1)
- [x] [collections-strawman](https://github.com/dotty-staging/collection-strawman/tree/sbt1)

### CODE REVIEW

**Comment by smarter** [comment]
Awesome! :)

**Comment by smarter** [comment]
@ttreyer Could you please sign the Scala CLA at https://www.lightbend.com/contribute/cla/scala ? That way we'll finally be able to merge your code :smile:

File: project/Build.scala [hunk 1]
```diff
@@ -150,8 +151,8 @@ object Build {
 
     // Avoid having to run `dotty-sbt-bridge/publishLocal` before compiling a bootstrapped project
     scalaCompilerBridgeSource :=
-      (dottyOrganization %% "dotty-sbt-bridge" % dottyVersion % Configurations.Component.name)
-      .artifacts(Artifact.sources("dotty-sbt-bridge").copy(url =
+      (dottyOrganization %% "dotty-sbt-bridge" % "NOT_PUBLISHED" % Configurations.Component.name)
```
Comment (smarter) [inline project/Build.scala]: A change was accidentally reverted here: we stopped using NOT_PUBLISHED in https://github.com/lampepfl/dotty/commit/78ac17ace84382322aadfa18334ba2a6b214d7ad

**Comment by smarter** [comment]
Until we're ready to merge this PR, we should enable running all the sbt scripted tests on every push to make sure everything passes.

**Comment by smarter** [comment]
https://github.com/sbt/zinc/pull/444 should be ported here too.

File: sbt-dotty/src/dotty/tools/sbtplugin/DottyPlugin.scala [hunk 2]
```diff
@@ -49,52 +51,43 @@ object DottyPlugin extends AutoPlugin {
...
-                  else "2.11"
-                case _ =>
-                  version
-              }
-            })
-          case _ =>
-            moduleID
-        }
-    }
+     implicit class DottyCompatModuleID(moduleID: ModuleID) {
```
Comment (smarter) [inline sbt-dotty/src/dotty/tools/sbtplugin/DottyPlugin.scala]: extra space in indentation

File: sbt-dotty/src/dotty/tools/sbtplugin/DottyPlugin.scala [hunk 3]
```diff
@@ -49,52 +51,43 @@ object DottyPlugin extends AutoPlugin {
...
+     }
   }
 
   import autoImport._
 
   override def requires: Plugins = plugins.JvmPlugin
   override def trigger = allRequirements
 
-  // Adapted from CrossVersionUtil#sbtApiVersion
+  // Adapted from CrossVersioconstant nUtil#sbtApiVersion
```
Comment (smarter) [inline sbt-dotty/src/dotty/tools/sbtplugin/DottyPlugin.scala]: typos

File: project/Build.scala [hunk 4]
```diff
@@ -194,35 +197,39 @@ object Build {
     libraryDependencies ++= {
       if (bootstrapFromPublishedJars.value)
         Seq(
-          dottyOrganization %% "dotty-library" % dottyNonBootstrappedVersion % Configurations.ScalaTool.name,
-          dottyOrganization %% "dotty-compiler" % dottyNonBootstrappedVersion % Configurations.ScalaTool.name
-        ).map(_.withDottyCompat())
+          dottyOrganization % "dotty-library_2.11" % dottyNonBootstrappedVersion % Configurations.ScalaTool.name,
+          dottyOrganization % "dotty-compiler_2.11" % dottyNonBootstrappedVersion % Configurations.ScalaTool.name
+        )//.map(_.withDottyCompat(scalaVersion.value))
```
Comment (smarter) [inline project/Build.scala]: This should be reverted now that `withDottyCompat` has been ported.

**Comment by Duhemm** [comment]
I addressed the comments so far. I've ported sbt/zinc#444 here.

All the scripted tests are already being run on every push for this PR.

I've also finished upgrading all the projects of the community build to sbt 1.0.3 and proposed to upstream the changes.

File: project/Build.scala [hunk 5]
```diff
@@ -853,21 +853,22 @@ object Build {
...
       sbtPlugin := true,
-      version := "0.1.7",
-      ScriptedPlugin.scriptedSettings,
-      ScriptedPlugin.sbtTestDirectory := baseDirectory.value / "sbt-test",
-      ScriptedPlugin.scriptedLaunchOpts += "-Dplugin.version=" + version.value,
-      ScriptedPlugin.scriptedLaunchOpts += "-Dplugin.scalaVersion=" + dottyVersion,
-     // By default scripted tests use $HOME/.ivy2 for the ivy cache. We need to override this value for the CI.
-      ScriptedPlugin.scriptedLaunchOpts ++= ivyPaths.value.ivyHome.map("-Dsbt.ivy.home=" + _.getAbsolutePath).toList,
-      ScriptedPlugin.scripted := ScriptedPlugin.scripted.dependsOn(Def.task {
+      version := "0.1.6-SNAPSHOT",
```
Comment (allanrenucci) [inline project/Build.scala]: 1.8

File: project/Build.scala [hunk 6]
```diff
@@ -199,7 +199,7 @@ object Build {
         Seq(
           dottyOrganization % "dotty-library_2.11" % dottyNonBootstrappedVersion % Configurations.ScalaTool.name,
           dottyOrganization % "dotty-compiler_2.11" % dottyNonBootstrappedVersion % Configurations.ScalaTool.name
```
Comment (smarter) [inline project/Build.scala]: This line and the line above also need to be reverted to not use "_2.11"

File: compiler/src/dotty/tools/dotc/sbt/ExtractAPI.scala [hunk 7]
```diff
@@ -254,8 +255,17 @@ private class ExtractAPICollector(implicit val ctx: Context) extends ThunkHolder
...
+        val ancestorTypes0 = linearizedAncestorTypes(cinfo)
+        if (ValueClasses.isDerivedValueClass(csym)) {
+          val underlying = ValueClasses.valueClassUnbox(csym).info.finalResultType
+          // The underlying type of a value class should be part of the name hash
+          // of the value class (see the test `value-class-underlying`), this is accomplished
+          // by adding the underlying type to the list of parent types.
+          underlying :: ancestorTypes0
+        } else
+          ancestorTypes0
+      } catch {
```
Comment (smarter) [inline compiler/src/dotty/tools/dotc/sbt/ExtractAPI.scala]: The try/catch should only be needed around the call to ` linearizedAncestorTypes(cinfo)`.

**Comment by Duhemm** [comment]
test performance please

**Comment by allanrenucci** [comment]
One more thing to do is to create a new docker image from this PR, update the image in `drone.yml` and make sure this PR does not introduce a big slow down on the tests.

File: project/Build.scala [hunk 8]
```diff
@@ -862,7 +862,7 @@ object Build {
       unmanagedSourceDirectories in Compile +=
         baseDirectory.value / "../language-server/src/dotty/tools/languageserver/config",
       sbtPlugin := true,
-      version := "0.1.6-SNAPSHOT",
+      version := "0.1.8",
```
Comment (smarter) [inline project/Build.scala]: This is an incompatible change, I would bump to 0.2.0

File: compiler/src/dotty/tools/dotc/sbt/ExtractDependencies.scala [hunk 9]
```diff
@@ -90,13 +99,59 @@ class ExtractDependencies extends Phase {
...
+    }
+  }
+
+  private def firstClassOrModule(tree: tpd.Tree)(implicit ctx: Context): Option[Symbol] = {
+    import tpd._
+    val acc = new TreeAccumulator[Option[Symbol]] {
+      def apply(x: Option[Symbol], t: Tree)(implicit ctx: Context) =
+        if (x.isDefined) x
+        else t match {
+          case moduleDef: Thicket =>
```
Comment (smarter) [inline compiler/src/dotty/tools/dotc/sbt/ExtractDependencies.scala]: I don't think this case is ever going to be hit, objects in Dotty after Typer are represented by a TypeDef and a ValDef, the TypeDef case below will be used.

File: compiler/src/dotty/tools/dotc/sbt/ExtractDependencies.scala [hunk 10]
```diff
@@ -90,13 +99,59 @@ class ExtractDependencies extends Phase {
...
+  }
+
+  private def classFile(sym: Symbol)(implicit ctx: Context): Option[AbstractFile] = {
+    // package can never have a corresponding class file; this test does not
+    // catch package objects (that do not have this flag set)
+    if (sym.is(Package)) None
+    else {
+      val file = Option(sym.associatedFile)
+
+      Option(sym.associatedFile).flatMap {
```
Comment (smarter) [inline compiler/src/dotty/tools/dotc/sbt/ExtractDependencies.scala]: I don't understand what this is for, if the symbol comes from source then `sym.associatedFile` will never be `NoSourceFile`.

File: compiler/src/dotty/tools/dotc/sbt/ExtractDependencies.scala [hunk 11]
```diff
@@ -105,34 +160,41 @@ class ExtractDependencies extends Phase {
...
-            // We can recover the fully qualified name of a classfile from
-            // its path
-            val classSegments = pf.givenPath.segments.takeRight(packages + 1)
-            binaryDependency(pf.file, className(classSegments))
-          case _ =>
-            ctx.warning(s"sbt-deps: Ignoring dependency $depFile of class ${depFile.getClass}")
-        }
-      } else if (depFile.file != currentSourceFile) {
-        ctx.sbtCallback.classDependency(extractedName(dep.enclosingClass), extractedName(currentClass), context)
+    val onSource = dep.sourceFile
```
Comment (smarter) [inline compiler/src/dotty/tools/dotc/sbt/ExtractDependencies.scala]: Why was this changed? This is also very likely to do the wrong thing. If `dep` was loaded from TASTY, `dep.sourceFile` will return the source filename stored in the SoureFile attribute of the TASTY section. But here we're really interested in recording the dependency on the classfile where `dep` comes from, because if this classfile changes we'll have to recompile the current unit.

File: compiler/src/dotty/tools/dotc/sbt/ExtractDependencies.scala [hunk 12]
```diff
@@ -161,11 +223,13 @@ private final class NameUsedInClass {
       scopedNames.get(name) match {
         case None =>
         case Some(otherScopes) =>
-          builder.append(" in [")
-          otherScopes.forEach(new java.util.function.Consumer[UseScope]() {
-            def accept(scope: UseScope): Unit =
+          // Pickling tests fail when this is turned in an anonymous class
```
Comment (smarter) [inline compiler/src/dotty/tools/dotc/sbt/ExtractDependencies.scala]: Could you open an issue so we don't forget about this?

File: compiler/src/dotty/tools/dotc/sbt/ExtractDependencies.scala [hunk 13]
```diff
@@ -182,13 +246,13 @@ private final class NameUsedInClass {
  *  specially, see the subsection "Dependencies introduced by member reference and
  *  inheritance" in the "Name hashing algorithm" section.
  */
-private class ExtractDependenciesCollector(implicit val ctx: Context) extends tpd.TreeTraverser { thisTreeTraverser =>
+private class ExtractDependenciesCollector(responsibleForImports: Symbol)(implicit val ctx: Context) extends tpd.TreeTraverser { thisTreeTraverser =>
```
Comment (smarter) [inline compiler/src/dotty/tools/dotc/sbt/ExtractDependencies.scala]: Please add some documentation for this parameter

File: compiler/src/dotty/tools/dotc/sbt/ExtractDependencies.scala [hunk 14]
```diff
@@ -199,15 +263,18 @@ private class ExtractDependenciesCollector(implicit val ctx: Context) extends tp
    *  because it refers to these classes or something defined in them.
    *  This is always a superset of `topLevelInheritanceDependencies` by definition.
    */
```
Comment (smarter) [inline compiler/src/dotty/tools/dotc/sbt/ExtractDependencies.scala]: Documentation needs to be updated since this is now a set of pair, same for the def just below.

File: compiler/src/dotty/tools/dotc/sbt/ExtractDependencies.scala [hunk 15]
```diff
@@ -199,15 +263,18 @@ private class ExtractDependenciesCollector(implicit val ctx: Context) extends tp
...
   /** The set of top-level classes that the compilation unit extends or that
    *  contain a non-top-level class that the compilaion unit extends.
    */
-  def topLevelInheritanceDependencies: Set[Symbol] = _topLevelInheritanceDependencies
+  def topLevelInheritanceDependencies: Set[(Symbol, Symbol)] = _topLevelInheritanceDependencies
 
   private def addUsedName(enclosingSym: Symbol, name: Name) = {
-    val enclosingName = extractedName(enclosingSym)
+    val enclosingName = enclosingSym match {
+      case sym if sym == defn.RootClass => ExtractDependencies.extractedName(responsibleForImports)
```
Comment (smarter) [inline compiler/src/dotty/tools/dotc/sbt/ExtractDependencies.scala]: Why use a pattern match and not just an if/else here?

File: compiler/src/dotty/tools/dotc/sbt/ExtractDependencies.scala [hunk 16]
```diff
@@ -276,19 +349,20 @@ private class ExtractDependenciesCollector(implicit val ctx: Context) extends tp
             addImported(name)
           case Thicket(Ident(name) :: Ident(rename) :: Nil) =>
             addImported(name)
-            if (rename ne nme.WILDCARD)
+            if (rename ne nme.WILDCARD) {
```
Comment (smarter) [inline compiler/src/dotty/tools/dotc/sbt/ExtractDependencies.scala]: Unnecessary brackets

File: compiler/src/dotty/tools/dotc/sbt/ExtractDependencies.scala [hunk 17]
```diff
@@ -241,10 +314,10 @@ private class ExtractDependenciesCollector(implicit val ctx: Context) extends tp
     sym.isAnonymousFunction ||
     sym.isAnonymousClass
 
-  private def addInheritanceDependency(sym: Symbol): Unit =
-    _topLevelInheritanceDependencies += sym.topLevelClass
+  private def addInheritanceDependency(sym: Symbol)(implicit ctx: Context): Unit =
+    _topLevelInheritanceDependencies += ((currentClass, sym.topLevelClass))
 
-  private object PatMatDependencyTraverser extends ExtractTypesCollector {
+  private class PatMatDependencyTraverser(ctx0: Context) extends ExtractTypesCollector(ctx0) {
```
Comment (smarter) [inline compiler/src/dotty/tools/dotc/sbt/ExtractDependencies.scala]: Why not `implicit ctx: Context` ? And at the same time, the `implicit ctx: Context` argument to method in this class should be removed.

File: compiler/src/dotty/tools/dotc/sbt/ThunkHolder.scala [hunk 18]
```diff
@@ -25,97 +25,8 @@ private[sbt] trait ThunkHolder {
...
-   *
-   *  TODO: Get rid of this method. It is only needed because some xsbti.api classes
-   *  take lazy arguments when they could be strict, but this can be fixed in sbt,
-   *  see https://github.com/sbt/zinc/issues/114
-   */
-  def strict2lzy[T <: AnyRef](t: T): api.Lazy[T] =
-    SafeLazyWrapper.strict(t)
-}
-
-/** Wrapper around SafeLazy implementations.
```
Comment (smarter) [inline compiler/src/dotty/tools/dotc/sbt/ThunkHolder.scala]: So great that we can get rid of this stuff :smile: :fire:

File: compiler/src/dotty/tools/dotc/sbt/ExtractAPI.scala [hunk 19]
```diff
@@ -182,32 +209,70 @@ private class ExtractAPICollector(implicit val ctx: Context) extends ThunkHolder
...
+        apiType(c.typeRef)
+      else
+        apiType(c.termRef)
+    ).toArray
+
+    val cl = api.ClassLike.of(
+      name, acc, modifiers, anns, defType, api.SafeLazy.strict(selfType), api.SafeLazy.strict(structure), Constants.emptyStringArray,
+      childrenOfSealedClass, topLevel, tparams)
+
+    // if (name.toString.contains("DottyPredef")) {
```
Comment (smarter) [inline compiler/src/dotty/tools/dotc/sbt/ExtractAPI.scala]: Remove debug code

File: compiler/src/dotty/tools/dotc/sbt/ExtractAPI.scala [hunk 20]
```diff
@@ -182,32 +209,70 @@ private class ExtractAPICollector(implicit val ctx: Context) extends ThunkHolder
...
+    //   ctx.atPhase(ctx.flattenPhase.next) { implicit ctx =>
+    //     println("flatten: " + sym.fullName.toString)
+    //     println("flattenm: " + sym.fullName.mangledString)
+    //   }
+    //   println("flattenx: " + toDenot(sym.binaryName.toString)
+    // }
+
+    allNonLocalClassesInSrc += cl
+
+    val javaPlatform = ctx.platform.asInstanceOf[JavaPlatform]
```
Comment (smarter) [inline compiler/src/dotty/tools/dotc/sbt/ExtractAPI.scala]: Please avoid the cast here, we don't want the compiler to crash when run with the Scala.js backend. I think we should just move `hasJavaMainMethod` outside of `JavaPlatform`

File: compiler/src/dotty/tools/dotc/sbt/ExtractDependencies.scala [hunk 21]
```diff
@@ -86,38 +160,84 @@ class ExtractDependencies extends Phase {
...
+  def extractedName(sym: Symbol)(implicit ctx: Context): String =
+    // ctx.atPhase(ctx.flattenPhase.next) { implicit ctx =>
+      if (sym.is(ModuleClass))
+        sym.fullName.stripModuleClassSuffix.toString
+      else
+        sym.fullName.toString
+    // }
+}
+
+private final class NameUsedInClass {
```
Comment (smarter) [inline compiler/src/dotty/tools/dotc/sbt/ExtractDependencies.scala]: All the added class/object in this PR should get a documentation comment.

File: compiler/src/dotty/tools/dotc/sbt/ExtractAPI.scala (hunk 20 above)
Comment (sjrd) [reply compiler/src/dotty/tools/dotc/sbt/ExtractAPI.scala]: Yes, especially since Scala.js supports exactly the same main methods as Scala/JVM!

**Comment by jvican** [comment]
Is there any chance this can be merged and released together with 0.5.0-RC1? I'd like to try cross-compiling bloop to Dotty. For now, I'm doing it locally, but ideally I'd have an official version to depend on so that the CI passes my changes.

**Comment by smarter** [comment]
> Is there any chance this can be merged and released together with 0.5.0-RC1?

No, that would require delaying the release too much. But once this PR is merged you'll be able to use our nightly builds in your project, using a nightly build of Dotty is just as easy as using a release.

**Comment by allanrenucci** [comment]
We had a regression with `sbt-pack` #3124: the generated jar contained the non-bootstrapped version. Since you are updating the version, it would be good to know if the issue is still present.
/cc @liufengyun

File: .drone.yml [hunk 22]
```diff
@@ -41,15 +41,25 @@ pipeline:
...
       # sbt scripted tests are slow and don't run on PRs
-      event: [ push, tag, deployment ]
+      # event: [ push, tag, deployment ]
+
+  test_sbt_2of2:
+    group: test
+    image: lampepfl/dotty:2017-10-20
+    commands:
+      - cp -R . /tmp/5/ && cd /tmp/5/
+      - ./project/scripts/sbt "sbt-dotty/scripted source-dependencies/*2of2"
```
Comment (allanrenucci) [inline .drone.yml]: It would be great if we did not introduce this huge regression (3x) in the time needed to run the scripted tests.
@Duhemm says it is because the bridge is recompiled for each scripted test. @smarter do you have an idea how this could be fixed?

**Comment by liufengyun** [comment]
Good catch @allanrenucci : `0.9.1` is the version where we had the regression. The latest version for `sbt-pack` is `1.0.4`. We can run `dist-bootstrapped/pack`, if the generated jars are good, then there's no worry.

File: .drone.yml (hunk 22 above)
Comment (smarter) [reply .drone.yml]: Complain to the sbt people: https://github.com/sbt/sbt/issues/3469 Maybe ask someone who has a lightbend subscription? :)

**Comment by allanrenucci** [comment]
Unfortunately `dist-bootstrapped/pack` generates the non bootstrapped jars

File: .drone.yml (hunk 22 above)
Comment (smarter) [reply .drone.yml]: Alternatively, we could try copy-pasting the custom implementation of scripted used by zinc, it even has parallelism and fancy colors contributed by @jvican :P. https://github.com/sbt/zinc/blob/1.x/project/Scripted.scala https://github.com/sbt/zinc/tree/1.x/internal/zinc-scripted/src/test/scala/sbt/internal/inc

**Comment by xerial** [comment]
@allanrenucci @liufengyun Released sbt-pack 0.10.0, which fixes an issue of including too many modules. https://github.com/xerial/sbt-pack/issues/67

**Comment by xerial** [comment]
Released sbt-pack 0.10.1, which fixes a bug if you have multiple nested projects in build.sbt. Please use this version. Thanks.

**Comment by allanrenucci** [comment]
Thanks @xerial! I can confirm this fixes our issue

**Comment by OlivierBlanvillain** [comment]
I'm closing this PR given that it's been rebased in #3687. @allanrenucci next time just merge master into the existing PR, rebasing doesn't scale to multiple committers and long standing PRs...
